pkp/pkp-lib#4915 display error if several blind participants are sele…

bozana opened the pull request:
…cted for a review discussion

s. https://github.com/pkp/pkp-lib/issues/4915

In the review stage also the guest editors, translators and funding coordinators can be assigned. In such a case it is then possible to select them in the review discussion participants list. Should they be considered and how -- are they allowed to see the participants names (e.g. like editors/section editors are)?

NateWr commented on controllers/grid/queries/form/QueryForm.inc.php:
```diff
@@ -323,6 +323,88 @@ function readInputData() {
+
+			// if current user is editor, forbid several blind participants (reviewers and author) in a review discussion
+			if ($user->hasRole([ROLE_ID_SITE_ADMIN], CONTEXT_SITE) ||
+				array_intersect([ROLE_ID_MANAGER, ROLE_ID_SUB_EDITOR], $assignedRoles)) {
```
I think this validation should occur always, not only when the user is an editor. Because a malicious user could still submit the form with any participant IDs, not just the ones that the UI shows to them.

NateWr commented on controllers/grid/queries/form/QueryForm.inc.php:
```diff
@@ -323,6 +323,88 @@ function readInputData() {
+			if ($user->hasRole([ROLE_ID_SITE_ADMIN], CONTEXT_SITE) ||
+				array_intersect([ROLE_ID_MANAGER, ROLE_ID_SUB_EDITOR], $assignedRoles)) {
+
+					if ($query->getStageId() == WORKFLOW_STAGE_ID_EXTERNAL_REVIEW || $query->getStageId() == WORKFLOW_STAGE_ID_INTERNAL_REVIEW) {
```
Likewise, I think we don't have to restrict this validation check to a particular stage. It should never be possible to open a discussion between anonymous reviewers and authors, right?

NateWr commented on controllers/grid/queries/form/QueryForm.inc.php:
```diff
@@ -323,6 +323,88 @@ function readInputData() {
+						if ($displayError) {
+							$this->addError('users', __('editor.review.discussion.errorAnonymousParticipants'));
+							$this->addErrorField('users');
+						}
```
Can I suggest a small adjustment that might take less logic? First, get a count of all authors and anonymous reviewers, then apply the validation logic. Like this:

```php
$participantIds = $this->getData('users');
$authorAssignments = $stageAssignmentDao->getBySubmissionAndRoleId($submissionId, ROLE_ID_AUTHOR);
$anonymousCount = 0;
while ($authorAssignment = $authorAssignments->next()) {
	if ($authorAssignment->getRoleId() == ROLE_ID_AUTHOR && in_array($authorAssignment->getUserId(), $participantIds) {
		$anonymousCount++;
	}
}
$reviewAssignments = $reviewAssignmentDao->getBySubmissionId($submissionId);
foreach ($reviewAssignments as $reviewAssignment) {
	if (in_array($authorAssignment->getUserId(), $participantIds)
			&& in_array($reviewAssignment->getReviewMethod(), [SUBMISSION_REVIEW_METHOD_ANONYMOUS, SUBMISSION_REVIEW_METHOD_DOUBLEANONYMOUS])) {
		$anonymousCount++;
	}
}
if ($anonymousCount > 1) {
	// apply error
}
```

bozana replied:
Here I am only not sure if several authors are selected and no reviewer -- they could see each other, or?
I would then also need to consider other roles (if this should apply regardless of stage and if stage participants are configurable), so there might be something like this then:
if participant role is not editor and not blind reviewer: count++
if participant is a blind reviewer: anonymousCount++
if (count > 0 && anonymousCount > 0) or (anonymousCount > 1): display error
I will then still have to see to do it with less logic, as you said... :-)
What do you think?

I will also have to see how to retrieve the participant roles...

NateWr replied:
> Here I am only not sure if several authors are selected and no reviewer -- they could see each other, or?

Ack! You're right! Either keep separate counts of the authors and reviewers. Or just break the `while` loop after the first author is found. That way it will only increment once if an author is in there.

> I would then also need to consider other roles

Ok, yeah, I guess if someone has an assistant or subeditor assignment to another stage but not the review stage, they shouldn't have access. I'm trying to think whether it would be possible to create such a discussion. The UI won't show them the participant IDs for selection when creating a discussion, but is it possible for a malicious user to add a participant who is not assigned to the stage? It seems like we can maybe tackle this problem from that direction: it shouldn't be possible to create a discussion in X stage with a participant who is not assigned to that stage.

NateWr commented on locale/en_US/editor.po:
```diff
@@ -272,6 +272,9 @@ msgstr "Unconsider this Review"
 msgstr "Once this review has been read, press \"Confirm\" to indicate that the review process may proceed. If the reviewer has submitted their review elsewhere, you may upload the file below and then press \"Confirm\" to proceed."
 
+msgid "editor.review.discussion.errorAnonymousParticipants"
+msgstr "You can not create a discussion that includes several anonymous reviewers, or the author and an anonymous reviewer, because the person's identity will be revealed."
```
Good! Tiny language adjustment:

> A discussion can not be created with the selected participants because it would impact the anonymity of the review process.

NateWr reviewed (commented):
Great! Just a few couple comments and one suggestion.

bozana commented:
@NateWr, I changed as discussed above, could you please take another look?
I also added a new error message -- if it happens that a user is selected that is not assigned to the workflow stage and is not a reviewer...
Thanks a lot!

NateWr commented on controllers/grid/queries/form/QueryForm.inc.php:
```diff
@@ -323,6 +323,65 @@ function readInputData() {
+	function validate($callHooks = true) {
+		// Display error if anonymity is impacted, i.e. if:
+		// 1) several blind reviewers are selected, or
+		// 2) a blind reviewer and another participant (other than editor) are selected.
```
I think it should be possible for an assistant assigned to the correct stage to see a blind reviewer...

bozana replied:
Is there a possibility to ensure that they are assigned to the correct stage (if it is configurable)? Or can I assume that this is a case and treat them as editors?

bozana replied:
Sorry for bothering you for one more thing: Actually, a participant can either be a reviewer or have another role for a submission and all stages -- i. e. a reviewer cannot be assigned e. g. as proofreader as well, or? -- I do not know if the system checks it, but it wouldn't make sense, or?
If that could be possible, currently (as I implemented it) the proofreader (that was also a reviewer) could not participate in a discussion together with e. g. author in the production stage. Correct so?

NateWr replied:
I guess if a user any stage assignment, they should be able to speak to authors. But only stage assignments in the same review stage should be able to speak to reviewers. (So in OMP we'll need to distinguish between internal/external review stages.)

Maybe the thing to do is to get the discussion's stage, then get a list of all stage assignments, reviewers and authors specific to that stage. And then to apply the logic which checks discussion assignments against these specific lists?

NateWr commented on controllers/grid/queries/form/QueryForm.inc.php:
```diff
@@ -323,6 +323,65 @@ function readInputData() {
+			$userGroupDao = DAORegistry::getDAO('UserGroupDAO'); /* @var $userGroupDao UserGroupDAO */
+
+			// get the selected participants
+			$newParticipantIds = $this->getData('users');
```
Maybe want to cast this to `array` just in case there are no users submitted and previous validation doesn't check that.

NateWr commented on controllers/grid/queries/form/QueryForm.inc.php:
```diff
@@ -323,6 +323,65 @@ function readInputData() {
+				// get participant review assignemnts
+				$reviewAssignments = $reviewAssignmentDao->getBySubmissionReviewer($submissionId, $participantId);
+				// if participant has no role in this stage and is not a reviewer, display error
+				if (empty($assignedRoles) && empty($reviewAssignments)) {
```
Will this prevent a journal manager who is not assigned at all to the submission from creating a discussion?

bozana replied:
Good catch!!! Thanks a lot!

NateWr reviewed (commented):
Looks good. A couple of queries but I think we're just about there!

bozana commented:
Sorry @NateWr, could you please take a look again?
-- admin and manager can now create a discussion and see everything
-- if a blind reviewer is also assigned as e.g. proofreader in the next workflow stage, there cannot be a discussion between him/her and author, because we do not differentiate stages. I suppose this is correct so? 
-- here https://github.com/pkp/pkp-lib/blob/master/controllers/grid/queries/form/QueryForm.inc.php#L234 manager that is not assigned is not considered, so if such a manager starts a review discussion he/she will not get the reviewers in the participants list to choose from. Is this correct?
Thanks a lot!!!

NateWr commented:
> I think now that we have the anonymity validation I can add the manager role to that admin role, so that such a manager can start a discussion with the reviewers as well?

Yes, I think that's right.

bozana commented:
@NateWr, maybe, hopefully a last look at the code?
-- I considered the manager role, as said above, so that also the manager that is not assigned sees the reviewers in the participants list
-- I am considering only the review stages for the anonymity validation -- reviewers are only active and visible in those stages
-- In all other stages we check if the participants are assigned in that stage. The assigned participants can see each other in those other stages.
Sorry for back and forth, and this and that... :-P

NateWr reviewed (approved):
Thanks @bozana! I think you've nailed it. :sweat_smile:

bozana commented:
Thanks a lot @NateWr!!! I'll then also make the submodule update PRs for OJS, OMP and OPS and if/when tests finish I will merge...
